service-providers/ip-packet-router/src/mixnet_listener.rs

```diff
@@ -48,46 +52,63 @@ impl ConnectedClients {
         let (connected_client_tx, connected_client_rx) = tokio::sync::mpsc::unbounded_channel();
         (
             Self {
-                clients: Default::default(),
+                clients_ipv4_mapping: Default::default(),
+                clients_ipv6_mapping: Default::default(),
                 tun_listener_connected_client_tx: connected_client_tx,
             },
             tun_listener::ConnectedClientsListener::new(connected_client_rx),
         )
     }
 
-    fn is_ip_connected(&self, ip: &IpAddr) -> bool {
-        self.clients.contains_key(ip)
+    fn is_ip_connected(&self, ips: &IPPair) -> bool {
+        self.clients_ipv4_mapping.contains_key(&ips.ipv4)
+            || self.clients_ipv6_mapping.contains_key(&ips.ipv6)
     }
 
-    fn get_client_from_ip_mut(&mut self, ip: &IpAddr) -> Option<&mut ConnectedClient> {
-        self.clients.get_mut(ip)
+    fn get_client_from_ip_mut(&mut self, ip: &IpAddr) -> Option<&mut ConnectedClientRef> {
+        match ip {
+            IpAddr::V4(ip) => self.clients_ipv4_mapping.get_mut(ip),
+            IpAddr::V6(ip) => self.clients_ipv6_mapping.get_mut(ip),
+        }
     }
 
-    fn is_nym_address_connected(&self, nym_address: &Recipient) -> bool {
-        self.clients
-            .values()
-            .any(|client| client.nym_address == *nym_address)
+    async fn is_nym_address_connected(&self, nym_address: &Recipient) -> bool {
```

Comment:
if we rely on this method in multiple places, it might become a bottleneck down the line since we're having to go through **all** connected clients and then wait trying to read **each** value. could we maybe have a reverse map somewhere? (not neccessarily now in this PR, but at some point in the future)

Reply:
Yep, you're right, this could be a bottleneck, and I think an unnecessary one too. I simplified the `Arc<RwLock>`, since they weren't needed everywhere anyway.